### ISSUE TITLE
Investigate frontend rework with React/Vue

### ISSUE THREAD

**Issue by DangerOnTheRanger** [DangerOnTheRanger/maniwani#77]
The point of this issue is to start a discussion on the matter of moving from using Jinja2 and Flask for Maniwani's frontend to one of React or Vue, the reason being that stuff like the live updating code requires duplication of template knowledge; posts are always rendered server-side, so as it stands the process for dynamically adding a new post looks something like this:
* Send post information back to server to be rendered there.
* Retrieve rendered HTML from server and render it on the client.
Dynamically-added new replies follow a similar process, but must duplicate some HTML templating since the server doesn't currently render replies for individual posts. It's all very ugly, and really needs to be streamlined.

Migrating the Jinja templates over to React/Vue would significantly streamline the whole process, as templates would be only stored once, and no information would need to be duplicated. To support clients who don't have Javascript enabled, the templates can be rendered server-side before sending the rendered result plus Javascript over the pipe.

The primary drawback of this approach is twofold, as I see it:
* The JS templates will need access to internal structures that ordinary API clients should not have access to (configuration, etc.) - in other words, there will need to be an internal backend API secured from access by the outside.
* Introducing JS-backed SSR will add another moving part into the mix, namely, a Node server running alongside Flask. While Docker can hide this somewhat, administratiors opting to take the non-Docker route will have to install and maintain Node in addition to the other Python dependencies. That said, Node is already a buildtime requirement, so it might not be too bad.

**Comment by DonaldTsang** [comment]
Vue > React for both speed and code readability.

**Comment by mission712** [comment]
I'd personally use a React SPA to implement this but it's up to you.

**Comment by DonaldTsang** [comment]
Do your own research but...
- https://github.com/krausest/js-framework-benchmark
- https://jsreport.io/js-frameworks-fastest/
- https://medium.com/@ajmeyghani/javascript-frameworks-performance-comparison-c566d19ab65b
- https://hackernoon.com/javascript-framework-comparison-with-examples-react-vue-hyperapp-97f064fb468d

**Comment by DangerOnTheRanger** [comment]
I think readability is a bit subjective; I find significant whitespace appealing, but others find it atrocious, for instance. As far as performance goes I've never seen anything that decisively proves that one front-end framework is faster than another - all the links I've seen (the above included) put React/Vue within a couple milliseconds of each other by most measures. I don't lean towards one framework or the other, but this would be a reasonably major undertaking, and some caution is due, I think.

@mission712 If you have any prior experiences with blazechan or just web development in general that might be relevant here, I would deeply appreciate hearing more about them - this is mostly outside my realm of expertise.

**Comment by DonaldTsang** [comment]
@mission712 Question: Why would a Single Page Application be a good idea? It may make the user experience feel smooth, but it would make archiving as difficult as it can come (unless you makeup a backup query string method or static HTML)?

@DangerOnTheRanger I would say that the main factor is whether you like Vue template syntax more, or you like JSX more. I will get some documents for your Django-oriented mind to decide.
- https://medium.freecodecamp.org/reacts-jsx-vs-vue-s-templates-a-showdown-on-the-front-end
- https://frontendmasters.com/courses/advanced-vue/jsx-vs-templates/
- https://gaya.pizza/articles/react-vs-vue-1-jsx-vue-template-syntax/
- https://medium.com/javascript-in-plain-english/i-created-the-exact-same-app-in-react-and-vue-here-are-the-differences

**Comment by mission712** [comment]
@DangerOnTheRanger I believe the main problem with most imageboard implementations is that the server rendered content is decoupled from the client rendered content (autoupdate posts, etc.). That is why I suggested a framework which uses a VDOM.
@DonaldTsang Server Side Rendering exists. I previously thought that doing everything on the frontend was a good idea but I have changed my mind. I think server side rendering gives everyone the best of both worlds, i.e. it will allow NoJS users to use the website while JS users will enjoy the full benefits of an SPA (page transition smoothness, more robust code, etc.) Something like Next.js or Nuxt.js using Maniwani strictly as an API server would be good.
